edb/schema/pointers.py

```diff
@@ -335,27 +380,26 @@ def get_derived(self, schema, source, target, *,
             ptr = schema.get(fqname, default=None)
             if ptr is None:
                 schema, ptr = self.derive_ref(
-                    schema, source, target,
+                    schema, source, target=target,
                     derived_name_base=derived_name_base, **kwargs)
-
+        assert isinstance(ptr, Pointer)
         return schema, ptr
 
     def get_derived_name_base(self, schema: s_schema.Schema) -> sn.Name:
         shortname = self.get_shortname(schema)
         return sn.Name(module='__', name=shortname.name)
 
     def derive_ref(
-        self: referencing.ReferencedT,
+        self,
         schema: s_schema.Schema,
-        source: s_sources.Source,
-        target: Optional[s_types.Type] = None,
+        referrer: so.QualifiedObject,
         *qualifiers: str,
+        target: Optional[s_types.Type] = None,
```

Comment:
To make the function signature compatible with the one defined in the ancestor `ReferencedObject`, I opted for moving the `target` argument among the keyword arguments. I think this is the most reasonable solution since we are using `**kwargs` here.